system_tests/program_gas_test.go

```diff
@@ -0,0 +1,437 @@
+package arbtest
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"math/big"
+	"regexp"
+	"strings"
+	"testing"
+
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/core/vm"
+	"github.com/ethereum/go-ethereum/eth/tracers/logger"
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/ethereum/go-ethereum/rpc"
+	"github.com/offchainlabs/nitro/arbos/util"
+	"github.com/offchainlabs/nitro/execution/gethexec"
+	"github.com/offchainlabs/nitro/solgen/go/mocksgen"
+	"github.com/offchainlabs/nitro/solgen/go/precompilesgen"
+	"github.com/offchainlabs/nitro/util/testhelpers"
+)
+
+func TestProgramSimpleCost(t *testing.T) {
+	ctx, l2info, l2client, auth := setupGasCostTest(t)
+	stylusProgram := deployWasm(t, ctx, auth, l2client, rustFile("hostio-test"))
+	evmProgram := deployEvmContract(t, ctx, auth, l2client, mocksgen.HostioTestMetaData)
+	otherProgram := deployWasm(t, ctx, auth, l2client, rustFile("storage"))
+	matchSnake := regexp.MustCompile("_[a-z]")
+
+	for _, tc := range []struct {
+		hostio  string
+		opcode  vm.OpCode
+		params  []any
+		maxDiff float64
+	}{
+		{hostio: "exit_early", opcode: vm.STOP},
+		{hostio: "transient_load_bytes32", opcode: vm.TLOAD, params: []any{common.HexToHash("dead")}},
+		{hostio: "transient_store_bytes32", opcode: vm.TSTORE, params: []any{common.HexToHash("dead"), common.HexToHash("beef")}},
+		{hostio: "return_data_size", opcode: vm.RETURNDATASIZE, maxDiff: 1.0},
+		{hostio: "account_balance", opcode: vm.BALANCE, params: []any{l2info.GetAddress("Owner")}},
+		{hostio: "account_code", opcode: vm.EXTCODECOPY, params: []any{otherProgram}},
+		{hostio: "account_code_size", opcode: vm.EXTCODESIZE, params: []any{otherProgram}},
+		{hostio: "account_codehash", opcode: vm.EXTCODEHASH, params: []any{otherProgram}},
+		{hostio: "evm_gas_left", opcode: vm.GAS, maxDiff: 1.0},
+		{hostio: "evm_ink_left", opcode: vm.GAS, maxDiff: 1.0},
+		{hostio: "block_basefee", opcode: vm.BASEFEE, maxDiff: 0.5},
+		{hostio: "chainid", opcode: vm.CHAINID, maxDiff: 1.0},
+		{hostio: "block_coinbase", opcode: vm.COINBASE, maxDiff: 0.5},
+		{hostio: "block_gas_limit", opcode: vm.GASLIMIT, maxDiff: 1.0},
+		{hostio: "block_number", opcode: vm.NUMBER, maxDiff: 1.0},
+		{hostio: "block_timestamp", opcode: vm.TIMESTAMP, maxDiff: 1.0},
+		{hostio: "contract_address", opcode: vm.ADDRESS, maxDiff: 0.5},
+		{hostio: "math_div", opcode: vm.DIV, params: []any{big.NewInt(1), big.NewInt(3)}},
+		{hostio: "math_mod", opcode: vm.MOD, params: []any{big.NewInt(1), big.NewInt(3)}},
+		{hostio: "math_pow", opcode: vm.EXP, params: []any{big.NewInt(1), new(big.Int).Lsh(big.NewInt(1), 255)}, maxDiff: 2}, // worst case
```

Comment:
maxDiff of 2 is pretty much checking nothing with current formula.
I think math_ functions will eventually be tested in our wasm gas test vs. execution time and not vs. EVM.
In the meantime, it's good to have them here, but if we don't have a good bound on math_pow vm it's better to compare it to a known constant than to EVM, with a comment explaining it.